src/dbnode/persist/fs/cross_block_reader.go

```diff
@@ -0,0 +1,274 @@
+	iOpts              instrument.Options
+}
+
+// NewCrossBlockReader constructs a new CrossBlockReader based on given DataFileSetReaders.
+// DataFileSetReaders must be configured to return the data in the order of index, and must be
+// provided in a slice sorted by block start time.
+// Callers are responsible for closing the DataFileSetReaders.
+func NewCrossBlockReader(dataFileSetReaders []DataFileSetReader, iOpts instrument.Options) (CrossBlockReader, error) {
+	var previousStart time.Time
+	for _, dataFileSetReader := range dataFileSetReaders {
+		if !dataFileSetReader.OrderedByIndex() {
+			return nil, errReaderNotOrderedByIndex
+		}
+		currentStart := dataFileSetReader.Range().Start
+		if !currentStart.After(previousStart) {
+			return nil, fmt.Errorf("dataFileSetReaders are not ordered by time (%s followed by %s)", previousStart, currentStart)
+		}
+		previousStart = currentStart
+	}
+
+	return &crossBlockReader{
+		dataFileSetReaders: append(make([]DataFileSetReader, 0, len(dataFileSetReaders)), dataFileSetReaders...),
+		records:            make([]BlockRecord, 0, len(dataFileSetReaders)),
+		iOpts:              iOpts,
+	}, nil
+}
+
+func (r *crossBlockReader) Next() bool {
+	if r.err != nil {
+		return false
+	}
+
+	var emptyRecord BlockRecord
+	if !r.started {
+		if r.err = r.start(); r.err != nil {
+			return false
+		}
+	} else {
+		// use empty var in inner loop with "for i := range" to have compiler use memclr optimization
+		// see: https://codereview.appspot.com/137880043
+		for i := range r.records {
+			r.records[i] = emptyRecord
+		}
+	}
+
+	if len(r.minHeap) == 0 {
+		return false
+	}
+
+	firstEntry, err := r.readOne()
+	if err != nil {
+		r.err = err
+		return false
+	}
+
+	r.id = firstEntry.id
+	r.tags = firstEntry.tags
+
+	r.records = r.records[:0]
+	r.records = append(r.records, BlockRecord{firstEntry.data, firstEntry.checksum})
+
+	for len(r.minHeap) > 0 && r.minHeap[0].id.Equal(firstEntry.id) {
+		nextEntry, err := r.readOne()
+		if err != nil {
+			// Close the resources that were already read but not returned to the consumer:
+			r.id.Finalize()
+			r.tags.Close()
+			for _, record := range r.records {
+				record.Data.DecRef()
+				record.Data.Finalize()
+			}
+			for i := range r.records {
+				r.records[i] = emptyRecord
+			}
+			r.records = r.records[:0]
+			r.err = err
+			return false
+		}
+
+		// id and tags not needed for subsequent blocks because they are the same as in the first block
+		nextEntry.id.Finalize()
+		nextEntry.tags.Close()
+
+		r.records = append(r.records, BlockRecord{nextEntry.data, nextEntry.checksum})
+	}
+
+	return true
+}
+
+func (r *crossBlockReader) Current() (ident.ID, ident.TagIterator, []BlockRecord) {
+	return r.id, r.tags, r.records
+}
+
+func (r *crossBlockReader) readOne() (*minHeapEntry, error) {
+	if len(r.minHeap) == 0 {
+		return nil, errEmptyReader
+	}
+
+	entry := heap.Pop(&r.minHeap).(*minHeapEntry)
+	if r.dataFileSetReaders[entry.dataFileSetReaderIndex] != nil {
```

Comment:
Should this be bounds checked? How much of a guarantee that this is always a valid index (no matter of what other state/lifecycle happens).

Reply:
On `start()`, we create `minHeap` entries with `dataFileSetReaderIndex` values in `range r.dataFileSetReaders`:
https://github.com/m3db/m3/pull/2617/files/718f086e3c13e8a4cc2715bc2fd89739e3ccaba1#diff-405ddf6e2ec332c15e3c63660abed14eR180-R181

From that point, whenever we consume `minHeap` entry, we read the next entry with the same `dataFileSetReaderIndex` value (from the same `DataFileSetReader`). Also `r.dataFileSetReaders` never shrinks.
So I'd say this is a really strong invariant, the logics would be completely broken without it.